### PR TITLE
feat(DMObjectBuilder): support late const definition

### LINKED ISSUE #803: 'Unknown identifier' with /var/const
Compiling /vg/station, an error occurs:
`Error at code\datums\supplypacks.dm:337:40: Unknown identifier "access_engine_minor"`
here:
```
330	/datum/supply_packs/sacidtank
331		name = "Sulphuric acid tank"
332		contains = list(/obj/structure/reagent_dispensers/sacid)
333		cost = 15
334		containertype = /obj/structure/closet/crate/secure/large/reinforced/shard/empty
335		containername = "sulphuric acid tank crate"
336		group = "Supplies"
337		one_access = list(access_engine_minor, access_science)
```
`access_engine_minor` is defined in `/code/game/jobs/access.dm` as:
`/var/const/access_engine_minor = 11	// Engineering Foyer`
This error only started occurring fairly recently.

### LINKED ISSUE #1011: Incorrect processing order of const object vars
It seems constant vars in objects aren't getting processed in the right order.  Perhaps related to #803

Test Case:
```
/datum/foo
  var/other_var = MY_CONSTANT
  var/const/MY_CONSTANT = 1
```

BYOND: Compilation Successful, behaves as expected

OpenDream: `Unknown identifier "MY_CONSTANT"`

### PR DESCRIPTION
**Pull request by Insineer**
Add support for late const definitions.

1. Add new `UnknownIdentifierException`
2. Put definition statements failed to compile with `UnknownIdentifierException` to "danglingStatements" object's dictionary
3. Process all statements failed because of unknown identifier again, when identifier is finally defined

fix #1011
fix #803

### CODE REVIEW

**Comment by Altoids1** [comment]
Does this handle the recursive case? Such as
```dm
var/const/A = B
var/const/B = C
var/const/C = A
```

**Comment by Insineer** [comment]
> Does this handle the recursive case? Such as
> 
> ```dm
> var/const/A = B
> var/const/B = C
> var/const/C = A
> ```

No, of course :)
BYOND can't compile this too:

```
code.dm:177:error: B:  compile failed (possible infinite cross-reference loop)
code.dm:178:error: C:  compile failed (possible infinite cross-reference loop)
code.dm:179:error: A:  compile failed (possible infinite cross-reference loop)
```

And that makes sense, cause which values A, B and C should take as a result?

**Comment by Altoids1** [comment]
It's fine that it doesn't emit as descriptive an error, I was just worried that the compiler would spin or something in such a situation.

A new unit test to show that we error in this situation would be a good addition to this PR :^)

**Comment by Insineer** [comment]
> A new unit test to show that we error in this situation would be a good addition to this PR :^)

Done

File: Content.Tests/DMProject/Tests/Const/Const_LateDefinition.dm [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+var/const/late_defined_const2 = late_defined_const
+var/const/late_defined_const = 1
+
+/obj/TestObj
+	var/a = 1 + const_var
+	var/const/const_var = 2
+
+/proc/RunTest()
+	var/obj/TestObj/o = new
+	return o.const_var + global_list[1]
```
Comment (wixoaGit) [inline Content.Tests/DMProject/Tests/Const/Const_LateDefinition.dm:13]: A test's return value isn't used for anything if there isn't a `// RETURN TRUE` comment at the top of the file. Instead, `ASSERT()` that the values are what they should be.
```suggestion
	ASSERT(o.const_var == 2)
	ASSERT(global_list[1] == 1)
```

File: DMCompiler/DM/DMObject.cs [hunk 2]
```diff
@@ -29,8 +30,8 @@ class DMObject {
 
         public bool IsRoot => Path == DreamPath.Root;
 
-        public List<DMASTObjectVarOverride>? danglingOverrides = null; // Overrides waiting for the LateVarDef event to happen
-
+        // Statements waiting for LateVarDef event to happen
+        public Dictionary<string, List<DMASTStatement>> danglingStatementsByUndefinedNames = new();
```
Comment (wixoaGit) [inline DMCompiler/DM/DMObject.cs:34]: Rename this to `DanglingStatementsByUndefinedNames` to fit the naming convention used throughout the rest of the project.

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: Content.Tests/DMProject/Tests/Const/Const_LateDefinition.dm (hunk 1 above)
Comment (Insineer) [reply Content.Tests/DMProject/Tests/Const/Const_LateDefinition.dm]: Done

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**